src/s3ql/backends/b2/b2_backend.py

```diff
@@ -0,0 +1,807 @@
+        }
+        response = self._do_api_call('b2_get_upload_url', request_data)
+
+        new_upload_url = urlparse(response['uploadUrl'])
+        new_authorization_token = response['authorizationToken']
+
+        upload_connection = HTTPConnection(new_upload_url.hostname, 443, ssl_context=self.ssl_context)
+        upload_connection.timeout = self.tcp_timeout
+
+        return {
+            'hostname': new_upload_url.hostname,
+            'connection': upload_connection,
+            'path': new_upload_url.path,
+            'authorizationToken': new_authorization_token,
+            'isUploading': False
+        }
+
+    def _invalidate_upload_url(self, upload_url_info):
+        '''Removes a no longer working upload url from the available
+        upload urls.
+        '''
+
+        log.debug('invalidating upload url: %s %s', upload_url_info['hostname'], upload_url_info['path'])
+
+        if upload_url_info['connection'] is not None:
+            upload_url_info['connection'].disconnect()
+
+        try:
+            self.available_upload_url_infos.remove(upload_url_info)
+        except ValueError:
+            pass
+
+        log.debug('upload urls left: %d', len(self.available_upload_url_infos))
+
+    @staticmethod
+    def _b2_url_encode(s):
+        '''URL-encodes a unicode string to be sent to B2 in an HTTP header.'''
+
+        s = str(s)
+        encoded_s = urllib.parse.quote(s.encode('utf-8'), safe='/\\')
+        encoded_s = encoded_s.replace('\\', '=5C')
+
+        return encoded_s
+
+    @staticmethod
+    def _b2_url_decode(s, decode_plus=True):
+        '''Decodes a Unicode string returned from B2 in an HTTP header.
+
+        Returns a Python unicode string.
+        '''
+
+        # Use str() to make sure that the input to unquote is a str, not
+        # unicode, which ensures that the result is a str, which allows
+        # the decoding to work properly.
+
+        s = str(s)
+        decoded_s = s.replace('=5C', '\\')
+        if decode_plus:
+            decoded_s = urllib.parse.unquote_plus(decoded_s)
+        else:
+            decoded_s = urllib.parse.unquote(decoded_s)
+        return decoded_s
+
+    def _create_metadata_dict(self, metadata, chunksize=2048):
+        metadata_dict = {}
+
+        info_header_count = 0
+        for key in metadata.keys():
+            if not isinstance(key, str):
+                raise ValueError('dict keys must be str, not %s' % type(key))
+
+            value = metadata[key]
+            if (not isinstance(value, (str, bytes, int, float, complex, bool))
+                and value is not None):
+                raise ValueError('value for key %s (%s) is not elementary' % (key, value))
+
+            if isinstance(value, (bytes, bytearray)):
+                value = base64.b64encode(value)
+
+            buffer = ('%s: %s' % (repr(key), repr(value)))
+            buffer = urllib.parse.quote(buffer, safe='!@#$^*()=+/?-_\'"><\\| `.,;:~')
```

Comment:
Why is this necessary? `_add_b2_metadata_to_headers` also calls `_b2_url_encode` on the value, doesn't it?
```suggestion
```